pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java

```diff
@@ -126,30 +143,73 @@ public int getNumUpdatesToZk() {
     return _numUpdatesToZk;
   }
 
+  @VisibleForTesting
+  TableRebalanceRetryConfig getTableRebalanceJobRetryConfig() {
+    return _tableRebalanceJobRetryConfig;
+  }
+
   private void trackStatsInZk() {
+    Map<String, String> jobMetadata =
+        createJobMetadata(_tableNameWithType, _rebalanceJobId, _tableRebalanceProgressStats,
+            _tableRebalanceJobRetryConfig);
+    _pinotHelixResourceManager.addControllerJobToZK(_rebalanceJobId, jobMetadata,
+        ZKMetadataProvider.constructPropertyStorePathForControllerJob(ControllerJobType.TABLE_REBALANCE),
+        prevJobMetadata -> {
+          // Abort the job when we're sure it has failed, otherwise continue to update the status.
+          if (prevJobMetadata == null) {
+            return true;
+          }
+          String prevStatsInStr = prevJobMetadata.get(RebalanceJobConstants.JOB_METADATA_KEY_REBALANCE_PROGRESS_STATS);
+          TableRebalanceProgressStats prevStats;
+          try {
+            prevStats = JsonUtils.stringToObject(prevStatsInStr, TableRebalanceProgressStats.class);
+          } catch (JsonProcessingException ignore) {
+            return true;
+          }
+          boolean jobFailed =
+              prevStats != null && RebalanceResult.Status.FAILED.toString().equals(prevStats.getStatus());
```

Comment:
Now I see how this abort is wired. I feel the abstraction is not correct though because basically we allow an observer to abort the execution by throwing an runtime exception. This is not safe because we cannot differentiate this abort vs other exceptions.
One option to make it cleaner is to add an API `isAborted()` into the observer, and call it in the `TableRebalancer`

Reply:
good point!